File: shared/typescript/retrieval.ts
/**
 * Retrieval primitives for lessons 34-38. Pure functions - no network.
 *
 * Voyage embeddings are L2-normalised, so a dot product ranks identically to
 * cosine similarity and is cheaper. `cosineSimilarity` is kept for vectors
 * from other providers that are not normalised.
 */

export function dotProduct(a: readonly number[], b: readonly number[]): number {
  if (a.length !== b.length) {
    throw new Error(`Dimension mismatch: ${a.length} vs ${b.length}`);
  }
  let sum = 0;
  for (let i = 0; i < a.length; i += 1) sum += a[i]! * b[i]!;
  return sum;
}

export function cosineSimilarity(a: readonly number[], b: readonly number[]): number {
  const norm = Math.sqrt(dotProduct(a, a)) * Math.sqrt(dotProduct(b, b));
  return norm === 0 ? 0 : dotProduct(a, b) / norm;
}

export interface Scored<T> {
  item: T;
  score: number;
}

export function topK<T>(scored: Scored<T>[], k: number): Scored<T>[] {
  return [...scored].sort((a, b) => b.score - a.score).slice(0, k);
}

/** Lowercase, strip punctuation, split on whitespace. */
export function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^\p{L}\p{N}\s]/gu, " ")
    .split(/\s+/)
    .filter((token) => token.length > 0);
}

/**
 * Okapi BM25 (lesson 37). Lexical retrieval that catches the exact
 * identifiers, product codes and proper nouns that embeddings blur.
 */
export class BM25Index<T> {
  private readonly docs: { item: T; tokens: string[]; length: number }[] = [];
  private readonly df = new Map<string, number>();
  private averageLength = 0;

  private readonly k1: number;
  private readonly b: number;

  // Written out rather than as constructor parameter properties: parameter
  // properties are not erasable syntax, so Node's built-in type stripping
  // (and `erasableSyntaxOnly` in tsconfig.json) rejects them.
  constructor(k1 = 1.5, b = 0.75) {
    this.k1 = k1;
    this.b = b;
  }

  add(item: T, text: string): void {
    const tokens = tokenize(text);
    this.docs.push({ item, tokens, length: tokens.length });
    for (const term of new Set(tokens)) {
      this.df.set(term, (this.df.get(term) ?? 0) + 1);
    }
    const total = this.docs.reduce((sum, doc) => sum + doc.length, 0);
    this.averageLength = total / this.docs.length;
  }

  search(query: string, k = 5): Scored<T>[] {
    const terms = tokenize(query);
    const scored = this.docs.map((doc) => {
      const counts = new Map<string, number>();
      for (const token of doc.tokens) {
        counts.set(token, (counts.get(token) ?? 0) + 1);
      }
      let score = 0;
      for (const term of terms) {
        const frequency = counts.get(term);
        if (!frequency) continue;
        const docFrequency = this.df.get(term) ?? 0;
        const idf = Math.log(
          1 + (this.docs.length - docFrequency + 0.5) / (docFrequency + 0.5),
        );
        const denominator =
          frequency +
          this.k1 * (1 - this.b + (this.b * doc.length) / this.averageLength);
        score += idf * ((frequency * (this.k1 + 1)) / denominator);
      }
      return { item: doc.item, score };
    });
    return topK(scored, k).filter((entry) => entry.score > 0);
  }

  get size(): number {
    return this.docs.length;
  }
}

/**
 * Reciprocal Rank Fusion (lesson 38): merge independent ranked lists without
 * needing their scores to share a scale.
 */
export function reciprocalRankFusion<T>(
  rankings: T[][],
  keyOf: (item: T) => string,
  k = 60,
): Scored<T>[] {
  const scores = new Map<string, number>();
  const items = new Map<string, T>();

  for (const ranking of rankings) {
    ranking.forEach((item, index) => {
      const key = keyOf(item);
      items.set(key, item);
      scores.set(key, (scores.get(key) ?? 0) + 1 / (k + index + 1));
    });
  }

  return [...scores.entries()]
    .map(([key, score]) => ({ item: items.get(key)!, score }))
    .sort((a, b) => b.score - a.score);
}

File: shared/python/course/retrieval.py
"""Retrieval primitives for lessons 34-38. Pure functions - no network.

Voyage embeddings are L2-normalised, so a dot product ranks identically to
cosine similarity and is cheaper. ``cosine_similarity`` is kept for vectors from
other providers that are not normalised.
"""

from __future__ import annotations

import math
import re
from collections import Counter
from dataclasses import dataclass
from typing import Generic, TypeVar

T = TypeVar("T")

_TOKEN_SPLIT = re.compile(r"[^\w]+", re.UNICODE)


def dot_product(a: list[float], b: list[float]) -> float:
    if len(a) != len(b):
        raise ValueError(f"Dimension mismatch: {len(a)} vs {len(b)}")
    return sum(x * y for x, y in zip(a, b, strict=True))


def cosine_similarity(a: list[float], b: list[float]) -> float:
    norm = math.sqrt(dot_product(a, a)) * math.sqrt(dot_product(b, b))
    return 0.0 if norm == 0 else dot_product(a, b) / norm


@dataclass(frozen=True)
class Scored(Generic[T]):
    item: T
    score: float


def top_k(scored: list[Scored[T]], k: int) -> list[Scored[T]]:
    return sorted(scored, key=lambda entry: entry.score, reverse=True)[:k]


def tokenize(text: str) -> list[str]:
    """Lowercase, strip punctuation, split on whitespace."""
    return [token for token in _TOKEN_SPLIT.split(text.lower()) if token]


class BM25Index(Generic[T]):
    """Okapi BM25 (lesson 37).

    Lexical retrieval that catches the exact identifiers, product codes and
    proper nouns that embeddings blur.
    """

    def __init__(self, k1: float = 1.5, b: float = 0.75) -> None:
        self.k1 = k1
        self.b = b
        self._docs: list[tuple[T, Counter[str], int]] = []
        self._df: Counter[str] = Counter()
        self._average_length = 0.0

    def add(self, item: T, text: str) -> None:
        tokens = tokenize(text)
        self._docs.append((item, Counter(tokens), len(tokens)))
        self._df.update(set(tokens))
        total = sum(length for _, _, length in self._docs)
        self._average_length = total / len(self._docs)

    def search(self, query: str, k: int = 5) -> list[Scored[T]]:
        terms = tokenize(query)
        scored: list[Scored[T]] = []
        for item, counts, length in self._docs:
            score = 0.0
            for term in terms:
                frequency = counts.get(term, 0)
                if not frequency:
                    continue
                df = self._df[term]
                idf = math.log(1 + (len(self._docs) - df + 0.5) / (df + 0.5))
                denominator = frequency + self.k1 * (
                    1 - self.b + self.b * length / self._average_length
                )
                score += idf * (frequency * (self.k1 + 1) / denominator)
            scored.append(Scored(item, score))
        return [entry for entry in top_k(scored, k) if entry.score > 0]

    def __len__(self) -> int:
        return len(self._docs)


def reciprocal_rank_fusion(
    rankings: list[list[T]],
    key_of,  # noqa: ANN001 - Callable[[T], str]
    k: int = 60,
) -> list[Scored[T]]:
    """Merge independent ranked lists without needing a shared score scale."""
    scores: dict[str, float] = {}
    items: dict[str, T] = {}
    for ranking in rankings:
        for index, item in enumerate(ranking):
            key = key_of(item)
            items[key] = item
            scores[key] = scores.get(key, 0.0) + 1 / (k + index + 1)
    return sorted(
        (Scored(items[key], score) for key, score in scores.items()),
        key=lambda entry: entry.score,
        reverse=True,
    )
